README
Publication Ingest

Python version used: 2.7.6

# Get weekly batch of papers from Thomson Reuters: The batch of papers
# is renewed every friday. So to get a weekly batch of papers, the
# papers should be downloaded every thursday and saved in the format of
# tr_mm_dd_yyyy_wk.bib where mm_dd_yyyy is the respective date for
# Monday in the next week (Monday when you are going to run the
# publication ingest)
# 
STEPS 1. Go to "http://isiknowledge.com/wos"

2. Go to "Advanced Search"

3. Paste the Query "AD=(University Florida OR Univ Florida OR UFL OR
UF)" in the search box. It will list the papers affiliated to the
University of Florida. The time range will determine the papers to be
downloaded.


4. Below the search box, there are the following options to give time
limits. -dropdown menu (gives the papers based on year, month, three
weeks or last week etc.) -from 'year' to 'year' (gives range from a
particular year to the desired year, for eg. 2008 to  2010)

 So now because the query is run on thursday, then the first drop down
 option is used which allows to list papers for latest week "Current
 week".


5. Hit search. It gives the batch of papers entered in Thomson Reuters
in the last week. Click on the batch of papers. Lists all the papers
with titles.


6. On top of the papers, the options "Send to" has a drop down menu arrow,
choose "other file formats". -Records :all records on page records from
__ to __ (Select this option and enter the number of papers to be
downloaded. For eg. if papers are 143, then range should be 1 to 143)
-Content : select "Full Record" -Format  : Choose Bibtex Hit "Send"


7. The file will be saved in the downloads as "savedrecs.bib". Rename the file as tr_mm_dd_yyyy_wk.bib. tr stands for
Thomson Reuters, mm for month, dd for the date and yyyy for year, wk is
for weekly (the date should be the monday when the ingest is to be
done). Save the file in the data subdirectory.


# Now, the first step is done. After you have the bibtex file, the file
# should be processed through various python files to get the final RDF,
# report and text file with all the titles added and respective DOIs.
# 
# Open the terminal window. (in mac: Launchpad>Utilities>Terminal ; in
# windows: open the start menu, type "cmd" and press enter, a terminal
# window opens). 

1. Run the shell script pub_ingest.sh. First the file will be processed by fix-bibtex.py, then by read-bibtex.py, and finally bibtex2.rdf will generate output RDF.


2. Open the report. There is a publisher report at the very top, and
tells how many new publishers were created. Check each of the publishers
created in the final report and search for the publisher(may be in
abbreviations) in the original bibtex file i.e. tr_mm_dd_yyyy_wk.bib.
The possible abbreviations are listed in the file vivotools.py.

3. Open the fix_bibtex.csv file with editor. It consists of all the
publishers with their abbreviations and corrected names.

#For eg.
#
    ADIS INT LTD|ADIS INTERNATIONAL LIMITED

The first part is the abbreviation of the publisher by which it was
added to VIVO initially. However, it is replaced by the correct and full
name.

In your report, find the abbreviated name of the publisher created in
the original bibtex file. Find the correct name of the publisher from
internet (Google etc.). In the fix_bibtex.csv, put the abbreviated name
and the corrected name of the publisher in the same format as above.
Repeat the process for all the publishers created (if the abbreviated
name is different from the correct name).

Save the file.

4. Re-run the process using the script (same previous commands: Step 1.
No need to change the pathname again).

You should have all the final five files now, output to the data subdirectory: tr_mm_dd_yyyy_wk.bib ;
tr_mm_dd_yyyy_wk_fin.bib ; tr_mm_dd_yyyy_wk_fin.rpt ;
tr_mm_dd_yyyy_wk_fin.lst ; tr_mm_dd_yyyy_wk_fin.rdf


5. Upload the RDF.

	a. Open "vivo.ufl.edu". Log into your account (You must have Editor
	rights to be able to upload RDF)

	b. On the top right corner of the screen, click on "Siteadmin".

	c. Siteadmin > Add/Remove RDF Data (Under "Advanced Data Tools")

	d. It gives you an option to choose a file to upload. Click on the
	option and choose the RDF "tr_mm_dd_yyyy_wk_fin.rdf

	e. Click "Submit"

	It takes 10-15 min to upload the RDF depending on the size of the
	file. When the upload is complete, the message will appear "RDF
	upload successful"



6. After the  upload is successful, the VIVO-Editor list should be
notified about the update with the umber of publications added and the
titles added.

	a. The mail should be sent to : " VIVO-EDITORS-L@lists.ufl.edu "
	with subject in format "'n' publications added to VIVO" (n stands
	for for the number of publications added)

	b. The exact number of publications added is given in the report as
	"publication created"

	b. The format of the mail should be:



		Hello,

		'n' publications have been added to VIVO. The publications are
		from the latest week batch. The titles are listed below.



		Regards,

		Your Name



		Copy paste the whole .lst file contents.

	d. Send out the mail.



#Disambiguation Cases: In the final report, at the very bottom, there is
#a whole list of the disambiguation cases. These are the cases where the
#publication added had an author similar to one existing in VIVO and the
#program could not identify the right author to attach the publication.
#Mostly, the links in disambiguation report are the stubs of a main
#profile. So, to remove these, you will need to merge the stubs to the
#main profile.
#
#
#
1. The report gives the URI where the publication is located. Below that
there is the name of the author and at least two URIs between which the
publication might be attached to.



2. Open the publication URI. Now open the author URIs listed below in
other tabs. There can be two cases:

	a. The URIs belong to the same person but are stubs (duplicate
	profiles which dont have UFID on them).In this case, find the main
	profile for the person and merge all the stubs to the main profile.
	For merging, follow the steps below:

		Siteadmin > Ingest Tools > Merge Resources

		A page opens allowing you to merge two different URIs. The first
		one is primary and below is secondary. Paste the main profile
		URI in the primary resource and URIs of the stub profiles in the
		secondary resource space one by one.

	b. The dismabiguation can be between two main profiles from
	different department. This will need your judgement regarding where
	and to which of the profiles, the publication should be linked to.
	This can be decided by matching the department of the person and the
	publication.

3. The process is repeated for every disambiguation case in the report.














CHANGELOG.md
    Version 1.0 MC 2012-08-16
    --  Using VIVO as the current state, process a bibtex file, checking VIVO
        for the presence of paper, authors, journal and publisher.  Create
        missing entities as needed.  Never delete or change anything.  Only
        add new items. Don't handle papers with more than 50 authors.

    Version 1.1 MC 2012-09-02
    --  Corporate authorship.  bibtex2rdf now handles author lists of any
        length. A parameter, MAX_AUTHORS controls how many authors appear in
        an authorship before the rest are placed in a corporate authorship.
    --  New report for disambiguation lists the paper and all the choices for
        an author that needs to be disambiguated
    --  Cosmetic improvements to paper titles -- roman numerals, 's, escaped
        chars &,<,>
    --  Fix bug in lst files that caused author lists to be out of order
    --  Fix bug in which bibtex with no journal field attempted to create
        publisher/journal cross links
    --  Version of the software now in the name of the file

      Version 1.2 MC 2012-11-25
    --  Use UF Entity rather than UFID as the base for the dictionaries to get
        more people into the dictionaries and cut down on stub creation
    --  Use key_string on people names to improve matches on mixed case names,
        names with space, punctuation and unicode
    --  Show the current time on the console at the start of each major step
        in the work
    --  Move make dictionary and find functions to vivotools

    Version 1.2.1 MC 2013-04-06
    --  Fix bug regarding disambiguated UF authors.  Version 1.2 erroneously
        created concatenated URIs in Authorships and authorInAuthorship. 1.2.1
        picks a URI.  This is consistent with previous versions.
    --  Updated harvestedBy text to indicate version 1.2.1
    Version 1.2.2 MC 2013-12-21
    --  Show name parts and number of publications for each alternative in
        the disambiguation report
    Version 1.3 MC 2014-01-01
    --  Uses update_pubmed to provide PMCID, NIHMSID, Abstract, full text link,
        MESH keywords and grants cited for all papers added that have a DOI and
        can be found in PubMed.  The Concept dictionary is created and
        updated as needed.
    --  Requires pybtex 1.6 which uses mixed case bibtex field names
    --  Provided test.bib for simple testing



fix_bibtex.py
#!/user/bin/env/python
"""
    fix_bibtex.py -- fix a bibtex file

    Thomson Reuters Bibtex has several errors that violate bibtex sytax and
    choke bibtex parsers.  These are:
      1.  Double brackets should be replaced with single brackets throughout.
          For example, Thomson Reuters often uses Year = {{1988}}.  Should be
          Year = {1988}
      2.  Bibtex keywords must not contain blanks.  Thomson Reuters asserts
          Web of Science Category = {...}  All assertions should be changed to
          Web-of-Science-Category = {...}
          Meeting Abstract should be changed to Meeting-Abstract = {...}
      3.  The values for Funding-Acknowledgement often (about 30% of the time)
          include a meaningless escape sequence \{[}\} which confuses the
          parser.  This string of six characters should be replaced everywhere
          with a single space.

    In addition, Thomson Reuters uses a series of abbreviations for journal
    names and publishers that can be improved on a case by case basis.

    This program reads a file of improvements, and a bibtex file from stdin,
    makes the improvements that need to be made, and writes an improved
    file to stdout

    Version 1.0 2012-08-25 MC
    --  Added additional publisher name corrections
    Version 1.1 2014-01-13 MC
    --  All data moved to a CSV file
    --  Conform with commenting and coding standards
"""
__author__ = "Michael Conlon"
__copyright__ = "Copyright 2014, University of Florida"
__license__ = "BSD 3-Clause license"
__version__ = "1.1"

import sys
import fileinput
from vivotools import read_csv
fix_bibtex = read_csv("fix_bibtex.csv")
for line in fileinput.input():
    for row in fix_bibtex.values():
        line = line.replace(row['original'], row['improved'])
    sys.stdout.write(line)



pub_ingest.py
#!/user/bin/env/python
"""
    pub_ingest.py -- Read a bibtex file and make VIVO RDF

    The following objects will be made as needed:
      -- publisher
      -- journal
      -- information resource
      -- timestamp for the information resource
      -- people
      -- authorships
      -- concepts

    The resulting ADD and SUB RDF file can then be read into VIVO

    To Do
    --  Complete refactor as an update process. Create resuable parts so that
        a publication can be created from bibtex, doi or pmid
    --  Improve DateTimeValue accuracy.  Currently all publications are entered
        as yearMonth precision.  Sometimes we have more information, sometimes
        we have less.  We should use the information as presented by the
        publisher, not overstate (yearMonth when there is only year) and not
        understate (yearMonth when we know the day).
    --  Reuse date objects -- only create dates when the appropriate date entity
        is not already in VIVO
    --  Update for VIVO-ISF
    --  Update or vivofoundation and vivopubs
"""

__author__ = "Michael Conlon"
__copyright__ = "Copyright 2014, University of Florida"
__license__ = "BSD 3-Clause license"
__version__ = "1.3"

import sys
from datetime import datetime, date
from pybtex.database.input import bibtex
import tempita
import vivotools

MAX_AUTHORS = 50

publisher_report = {}
journal_report = {}
title_report = {}
author_report = {}
disambiguation_report = {}

dictionaries = []
journal_dictionary = {}
publisher_dictionary = {}
title_dictionary = {}

def open_files(bibtex_file_name):
    """
    Give the name of the bibitex file to be used as input, generate the file
    names for rdf, rpt and lst.  Return the open file handles
    """
    base = bibtex_file_name[:bibtex_file_name.find('.')]
    rpt_file = open(base+'.rpt', 'w')
    lst_file = open(base+'.lst', 'w')
    rdf_file = open(base+'.rdf', 'w')
    return [rdf_file, rpt_file, lst_file]

def update_disambiguation_report(authors, publication_uri):
    """
    Given the authors structure and thte publication_uri, add to the report
    if any of the authors need to be disambiguated
    """
    for value in authors.values():
        if value[8] == "Disambig":
            if publication_uri in disambiguation_report:
                result = disambiguation_report[publication_uri]
                result[len(result.keys())+1] = value
                disambiguation_report[publication_uri] = result
            else:
                disambiguation_report[publication_uri] = {1:value}
    return

# start here.  Create a parser for bibtex and use it to read the file of
# bibtex entries. open the output files

print datetime.now(), "Read the BibTex"
bibtex_file_name = sys.argv[1]
[rdf_file, rpt_file, lst_file] = open_files(bibtex_file_name)
parser = bibtex.Parser()
bib_data = parser.parse_file(bibtex_file_name)
bib_sorted = sorted(bib_data.entries.items(),
    key=lambda x: x[1].fields['title'])

print >>rdf_file, "<!--", len(bib_data.entries.keys()),\
    "publications to be processed -->"
print datetime.now(), len(bib_data.entries.keys()),\
    "publications to be processed."

#  make dictionaries for people, papers, publishers, journals, concepts

print datetime.now(), "Creating the dictionaries"
print datetime.now(), "Publishers"
publisher_dictionary = vivotools.make_publisher_dictionary()
print datetime.now(), "Journals"
journal_dictionary = vivotools.make_journal_dictionary()
print datetime.now(), "People"
dictionaries = make_people_dictionaries()
print datetime.now(), "Titles"
title_dictionary = vivotools.make_title_dictionary()
print datetime.now(), "Concepts"
vivotools.make_concept_dictionary()

# process the papers

print >>rdf_file, vivotools.rdf_header()

for key, value in bib_sorted:
    try:
        title = value.fields['title'].title() + " "
    except:
        title_report["No title"] = ["No Title", None, 1]
        print >>rdf_file, "<!-- No title found. No RDF necessary -->"
        continue
    title = abbrev_to_words(title)
    title = title[0:-1]
    if title in title_report:
        print >>rdf_file, "<!-- Title", title,\
            "handled previously.  No RDF necessary -->"
        title_report[title][2] = title_report[title][2] + 1
        continue
    else:
        print >>rdf_file, "<!-- Begin RDF for " + title + " -->"
        print datetime.now(), "<!-- Begin RDF for " + title + " -->"
        document = {}
        document['title'] = title
        title_report[title] = ["Start", None, 1]
        [found, uri] = vivotools.find_title(title, title_dictionary)
        if not found:
            title_report[title][0] = "Create" # Create

            # Authors

            [author_rdf, authors] = make_author_rdf(value)
            document['authors'] = make_document_authors(authors)
            if count_uf_authors(authors) == 0:
                print >>rdf_file, "<!-- End RDF.  No UF authors for " +\
                    title + " No RDF necessary -->"
                title_report[title][0] = "No UF Auth"
                continue
            update_author_report(authors)

            # Datetime

            [datetime_rdf, datetime_uri] = make_datetime_rdf(value, title)

            # Publisher

            [journal_create, journal_name, journal_uri] =\
                make_journal_uri(value)
            [publisher_create, publisher, publisher_uri, publisher_rdf] =\
                make_publisher_rdf(value)

            # Journal

            [journal_rdf, journal_uri] = make_journal_rdf(value,\
                journal_create, journal_name, journal_uri)

            # Publisher/Journal bi-directional links

            publisher_journal_rdf = ""
            if journal_uri != "" and publisher_uri != "" and\
                (journal_create or publisher_create):
                publisher_journal_rdf = \
                    make_publisher_journal_rdf(publisher_uri, journal_uri)

            # Authorships

            publication_uri = vivotools.get_vivo_uri()
            title_report[title][1] = publication_uri 
            [authorship_rdf, authorship_uris] = make_authorship_rdf(authors,\
                publication_uri)

            # AuthorInAuthorships

            author_in_authorship_rdf = make_author_in_authorship_rdf(authors,\
                authorship_uris)

            #  Journal/Publication bi-directional links

            if journal_uri != "" and publication_uri != "":
                journal_publication_rdf = \
                    make_journal_publication_rdf(journal_uri, publication_uri)

            #  PubMed values

            pubmed_rdf = ""
            if 'doi' in value.fields:
                [pubmed_rdf, sub] = vivotools.update_pubmed(publication_uri,\
                    value.fields['doi'])
                if sub != "":
                    raise Exception("Non empty subtraction RDF"+\
                        "for Update PubMed")

            # Publication

            publication_rdf = make_publication_rdf(value,\
                title,publication_uri,datetime_uri,authorship_uris)
            print >>rdf_file, datetime_rdf, publisher_rdf, journal_rdf,\
                publisher_journal_rdf, author_rdf, authorship_rdf,\
                author_in_authorship_rdf, journal_publication_rdf,\
                publication_rdf, pubmed_rdf
            print >>rdf_file, "<!-- End RDF for " + title + " -->"
            print >>lst_file, vivotools.string_from_document(document),\
                'VIVO uri', publication_uri, '\n'
            update_disambiguation_report(authors, publication_uri)
        else:
            title_report[title][0] = "Found"
            title_report[title][1] = uri
            print >>rdf_file, "<!-- Found: " + title + " No RDF necessary -->"
print >>rdf_file, vivotools.rdf_footer()

#
# Reports
#
print >>rpt_file,"""

Publisher Report

Lists the publishers that appear in the bibtex file in alphabetical order.  For
each publisher, show the improved name, the number of papers in journals of this publisher,
the action to be taken for the publisher and the VIVO URI -- the URI is the new
URI to be created if Action is Create, otherwise it is the URI of the found publisher
in VIVO.

Publisher                             Papers Action VIVO URI
---------------------------------------------------------------------------------"""
publisher_count = 0
actions = {}
for publisher in sorted(publisher_report.keys()):
    publisher_count = publisher_count + 1
    [create,uri,count] = publisher_report[publisher]
    if create:
        result = "Create"
    else:
        result = "Found "
    actions[result] = actions.get(result,0) + 1
    print >>rpt_file, "{0:40}".format(publisher[0:40]),"{0:>3}".format(count),result,uri
print >>rpt_file,""
print >>rpt_file, "Publisher count by action"
print >>rpt_file, ""
for action in sorted(actions):
    print >>rpt_file, action,actions[action]
print >>rpt_file, publisher_count,"publisher(s)"

print >>rpt_file, """

Journal Report

Lists the journals that appear in the bibtex file in alphabetical order.  For
each journal, show the improved name, the number of papers t be linked to the journal,
the action to be taken for the journal and the VIVO URI -- the URI is the new
URI to be created if Action is Create, otherwise it is the URI of the found journal
in VIVO.

Journal                               Papers Action VIVO URI
---------------------------------------------------------------------------------"""
journal_count = 0
actions = {}
for journal in sorted(journal_report.keys()):
    journal_count = journal_count + 1
    [create,uri,count] = journal_report[journal]
    if create:
        result = "Create"
    else:
        result = "Found "
    actions[result] = actions.get(result,0) + 1
    print >>rpt_file, "{0:40}".format(journal[0:40]),"{0:>3}".format(count),result,uri
print >>rpt_file, ""
print >>rpt_file, "Journal count by action"
print >>rpt_file, ""
for action in sorted(actions):
    print >>rpt_file, action,actions[action]
print >>rpt_file, journal_count,"journal(s)"


print >>rpt_file, """

Title Report

Lists the titles that appear in the bibtex file in alphabetical order.  For
each title, show the action to be taken, the number of times the title appears in
the bibtex, the improved title and the VIVO URI of the publication -- the URI is the new
URI to be created if action is Create, otherwise it is the URI of the found publication
in VIVO.

Action   # Title and VIVO URI
---------------------------------------------------------------------------------"""
title_count = 0
actions = {}
for title in sorted(title_report.keys()):
    title_count = title_count +1
    [action,uri,count] = title_report[title]
    actions[action] = actions.get(action,0) + 1
    print >>rpt_file, "{0:>10}".format(action),title,uri
print >>rpt_file, ""
print >>rpt_file, "Title count by action"
print >>rpt_file, ""
for action in sorted(actions):
    print >>rpt_file, action,actions[action]
print >>rpt_file, title_count,"title(s)"

print >>rpt_file, """

Author Report

For each author found in the bibtex file, show the author's name followed by the number of papers
for the author in the bibtex to be entered, followed by 
a pair of results for each time the author appears on a paper in the bibtex.  The result
pair contains an action and a URI.  The action is "non UF" if a non-UF author stub will be
be created, the URI is the URI of the new author stub.  Action "Make UF" if a new UF author
stub will be created with the URI of the new author stub.  "Found UF" indicate the author was
found at the URI.  "Disambig" if multiple UF people were found with the given name.  The URI
is the URI of one of the found people.  Follow-up is needed to determine if correct and
reassign author if not correct.

Author                    Action   URI                                          Action   URI
----------------------------------------------------------------------------------------------"""
author_count = 0
actions = {}
for author in sorted(author_report.keys()):
    author_count = author_count + 1
    results = ""
    papers = len(author_report[author])
    action = author_report[author][1][8] # 1st report, 8th value is action
    actions[action] = actions.get(action,0) + 1    
    for key in author_report[author].keys():
        value = author_report[author][key]
        results = results + value[8] + " " + "{0:45}".format(value[9])
    print >>rpt_file, "{0:25}".format(author),"{0:>3}".format(papers),results
print >>rpt_file, ""
print >>rpt_file, "Author count by action"
print >>rpt_file, ""
for action in sorted(actions):
    print >>rpt_file, action,actions[action]
print >>rpt_file, author_count,"authors(s)"

print >>rpt_file, """

Disambiguation Report

For each publication with one or more authors to disambiguate, list the paper, and
then the authors in question with each of the possible URIs to be disambiguated, show the URI
of the paper, and then for each author that needs to be disambiguated on the paper, show
the last name, first name and middle initial and the all the URIs in VIVO for UF persons
with the same names.
"""

for uri in disambiguation_report.keys():
    print >>rpt_file,"The publication at",uri,"has one or more authors in question"
    for key,value in disambiguation_report[uri].items():
        uris = value[9].split(";")
        print >>rpt_file,"    ",value[4],value[5],value[6],":"
        for u in uris:
            person = vivotools.get_person(u)
            if 'last_name' not in person:
                person['last_name'] = "No last name"
            if 'middle_name' not in person:
                person['middle_name'] = "No middle name"
            if 'first_name' not in person:
                person['first_name'] = "No first name"
            if 'home_department_name' not in person:
                person['home_department_name'] = "No home department"
            npubs = len(person['authorship_uris'])
            print >>rpt_file,"        ",u,person['last_name'], \
                person['first_name'],person['middle_name'], \
                person['home_department_name'],"Number of pubs = ",npubs
        print >>rpt_file
    print >>rpt_file
#
#  Close the files, we're done
#
rpt_file.close()
rdf_file.close()
lst_file.close()

    
